File: spikes/todo/src/components/TodoList.js
import React, {useContext} from 'react'
import Todo from './Todo'
import TodosContext from '../context/TodosContext'
import {VisibilityFilters} from "../constants/visibilityFilters";
import VisibilityFilterContext from "../context/VisibilityFilterContext";

const TodoList = () => {
    const todosContext = useContext(TodosContext)
    const visibilityFilterContext = useContext(VisibilityFilterContext)

    return <ul>
        {
            getVisibleTodos(todosContext.todos, visibilityFilterContext.visibilityFilter).map(todo => (
                <Todo key={todo.id} {...todo} onClick={() => todosContext.toggleTodo(todo.id)}/>
            ))
        }
    </ul>
}

const getVisibleTodos = (todos, filter) => {
    console.log(`getVisibleTodos ${filter}`)
    switch (filter) {
        case VisibilityFilters.SHOW_COMPLETED:
            return todos.filter(t => t.completed)
        case VisibilityFilters.SHOW_ACTIVE:
            return todos.filter(t => !t.completed)
        default:
            return todos
    }
}

export default TodoList

File: ui-spikes/src/main/scala/ui/laminar/models/TodoList.scala
package ui.laminar.models

case class TodoList(todos: Vector[Todo]) {
  def add(text: String): TodoList = TodoList {
    todos :+ Todo(todos.length, text, isComplete = false)
  }

  def toggle(id: Int) = TodoList {
    todos.map { todo =>
      if (todo.id == id) todo.copy(isComplete = !todo.isComplete) else todo
    }
  }

  def filter(visibilityFilter: VisibilityFilter): TodoList = TodoList {
    visibilityFilter match {
      case VisibilityFilter.All       => todos
      case VisibilityFilter.Completed => todos.filter(_.isComplete)
      case VisibilityFilter.Active    => todos.filter(!_.isComplete)
    }
  }
}
